File: ruby/lib/epostak/resources/webhook_queue.rb
# frozen_string_literal: true

require "erb"

module EPostak
  module Resources
    # Sub-resource for the webhook pull queue -- an alternative to push webhooks.
    # Use the pull queue when your server cannot receive inbound HTTPS requests.
    # Events accumulate in the queue and must be acknowledged after processing.
    #
    # Accessed via +client.webhooks.queue+.
    #
    # @example Poll-based event consumption loop
    #   loop do
    #     response = client.webhooks.queue.pull(limit: 50)
    #     break if response["events"].empty?
    #
    #     ids = response["events"].map { |item| process(item); item["event_id"] }
    #     client.webhooks.queue.batch_ack(ids)
    #     break if response["count"] == 0
    #   end
    class WebhookQueue
      # @param http [EPostak::HttpClient] Internal HTTP client
      def initialize(http)
        @http = http
      end

      # Pull unacknowledged events from the webhook queue.
      # Events remain in the queue until explicitly acknowledged via +ack+ or +batch_ack+.
      #
      # @param limit [Integer, nil] Maximum number of events to return
      # @param event_type [String, nil] Filter by event type (e.g. "document.received")
      # @return [Hash] Response with "events" array (each has event_id, firm_id, event, created_at, payload) and "count"
      #
      # @example
      #   response = client.webhooks.queue.pull(limit: 20, event_type: "document.received")
      #   response["events"].each { |item| puts item["event"] }
      def pull(limit: nil, event_type: nil)
        query = { limit: limit, event_type: event_type }
        @http.request(:get, "/webhook-queue", query: query)
      end

      # Acknowledge (remove) a single event from the queue after processing.
      #
      # @param event_id [String] The event ID to acknowledge
      # @return [Hash] { "acknowledged" => true }
      #
      # @example
      #   result = client.webhooks.queue.ack("event-uuid")
      #   puts result["acknowledged"] # => true
      def ack(event_id)
        @http.request(:delete, "/webhook-queue/#{encode(event_id)}")
      end

      # Acknowledge (remove) multiple events from the queue in a single request.
      #
      # @param event_ids [Array<String>] Array of event IDs to acknowledge
      # @return [Hash] { "acknowledged" => N } with count of acknowledged events
      #
      # @example
      #   response = client.webhooks.queue.pull(limit: 50)
      #   ids = response["events"].map { |i| i["event_id"] }
      #   result = client.webhooks.queue.batch_ack(ids)
      #   puts "Acknowledged: #{result['acknowledged']}"
      def batch_ack(event_ids)
        @http.request(:post, "/webhook-queue/batch-ack", body: { event_ids: event_ids })
      end

      # Pull events across all managed firms (integrator endpoint).
      # Only available with integrator API keys (+sk_int_*+).
      # Use the +since+ parameter for cursor-based polling.
      #
      # @param limit [Integer, nil] Maximum number of events to return
      # @param since [String, nil] ISO 8601 timestamp for cursor-based polling
      # @return [Hash] Response with "events" array and "count"
      #
      # @example
      #   response = client.webhooks.queue.pull_all(since: "2026-04-11T00:00:00Z", limit: 200)
      #   response["events"].each { |e| puts "#{e['firm_id']}: #{e['type']}" }
      def pull_all(limit: nil, since: nil)
        query = { limit: limit, since: since }
        @http.request(:get, "/webhook-queue/all", query: query)
      end

      # Acknowledge (remove) multiple events from the cross-firm queue (integrator endpoint).
      # Only available with integrator API keys (+sk_int_*+).
      #
      # @param event_ids [Array<String>] Array of event IDs to acknowledge
      # @return [Hash] Response with "acknowledged" count
      #
      # @example
      #   response = client.webhooks.queue.pull_all(limit: 100)
      #   ids = response["events"].map { |e| e["event_id"] }
      #   result = client.webhooks.queue.batch_ack_all(ids)
      #   puts "Acknowledged: #{result['acknowledged']}"
      def batch_ack_all(event_ids)
        @http.request(:post, "/webhook-queue/all/batch-ack", body: { event_ids: event_ids })
      end

      private

      def encode(value)
        ERB::Util.url_encode(value)
      end
    end
  end
end

File: php/src/Resources/WebhookQueue.php
<?php

declare(strict_types=1);

namespace EPostak\Resources;

use EPostak\HttpClient;
use EPostak\EPostakError;

/**
 * Pull-based webhook event queue.
 *
 * An alternative to push webhooks: poll for pending events and acknowledge
 * them after processing. Access via `$client->webhooks->queue`.
 */
class WebhookQueue
{
    /**
     * @param HttpClient $http Shared HTTP transport instance.
     */
    public function __construct(private HttpClient $http)
    {
    }

    /**
     * Pull pending webhook events from the queue.
     *
     * Events must be acknowledged after processing via ack() or batchAck().
     * Unacknowledged events will be returned again on the next pull.
     *
     * @param array{limit?: int, event_type?: string} $params Optional filters:
     *   - `limit`      Max events to return (default 20).
     *   - `event_type` Filter to a specific event type (e.g. 'document.received').
     * @return array{items: array, has_more: bool} Pending events and pagination flag.
     * @throws EPostakError On API error.
     *
     * @example
     *   $result = $client->webhooks->queue->pull(['limit' => 50]);
     *   foreach ($result['items'] as $item) {
     *       processEvent($item);
     *       $client->webhooks->queue->ack($item['event_id']);
     *   }
     */
    public function pull(array $params = []): array
    {
        $qs = HttpClient::buildQuery([
            'limit' => $params['limit'] ?? null,
            'event_type' => $params['event_type'] ?? null,
        ]);
        return $this->http->request('GET', '/webhook-queue' . $qs);
    }

    /**
     * Acknowledge a single event, removing it from the queue.
     *
     * @param string $eventId Event UUID to acknowledge.
     * @return array{acknowledged: true} Acknowledgement confirmation.
     * @throws EPostakError On API error.
     */
    public function ack(string $eventId): array
    {
        return $this->http->request('DELETE', '/webhook-queue/' . urlencode($eventId));
    }

    /**
     * Batch acknowledge multiple events, removing them from the queue.
     *
     * Max 1000 event IDs per call.
     *
     * @param string[] $eventIds Array of event UUIDs to acknowledge (max 1000).
     * @return array{acknowledged: int} Count of acknowledged events.
     * @throws EPostakError On API error.
     */
    public function batchAck(array $eventIds): array
    {
        return $this->http->request('POST', '/webhook-queue/batch-ack', [
            'json' => ['event_ids' => $eventIds],
        ]);
    }

    /**
     * Pull events across all firms (integrator only).
     *
     * Requires an integrator API key. Returns events from all assigned firms.
     *
     * @param array{limit?: int, since?: string} $params Optional filters:
     *   - `limit` Max events to return (default 100, max 500).
     *   - `since` ISO 8601 datetime cutoff.
     * @return array{items: list<array{event_id: string, firm_id: string, event: string, payload: array, created_at: string}>, has_more: bool}
     *   Pending events from all firms; `has_more` indicates more events are available.
     * @throws EPostakError On API error.
     */
    public function pullAll(array $params = []): array
    {
        $qs = HttpClient::buildQuery([
            'limit' => $params['limit'] ?? null,
            'since' => $params['since'] ?? null,
        ]);
        return $this->http->request('GET', '/webhook-queue/all' . $qs);
    }

    /**
     * Batch acknowledge events across all firms (integrator only).
     *
     * Requires an integrator API key. Max 1000 event IDs per call.
     *
     * @param string[] $eventIds Array of event UUIDs to acknowledge (max 1000).
     * @return array{acknowledged: int} Count of acknowledged events.
     * @throws EPostakError On API error.
     */
    public function batchAckAll(array $eventIds): array
    {
        return $this->http->request('POST', '/webhook-queue/all/batch-ack', [
            'json' => ['event_ids' => $eventIds],
        ]);
    }
}
